File: src/main/java/org/irislang/jiris/core/IrisClosureBlock.java
package org.irislang.jiris.core;

import org.irislang.jiris.compiler.IrisInterpreter;
import org.irislang.jiris.core.exceptions.IrisExceptionBase;
import org.irislang.jiris.core.exceptions.fatal.IrisConstanceNotFoundException;
import org.irislang.jiris.core.exceptions.fatal.IrisParameterNotFitException;
import org.irislang.jiris.core.exceptions.fatal.IrisUnkownFatalException;
import org.irislang.jiris.dev.IrisDevUtil;

import java.lang.invoke.MethodHandle;
import java.util.ArrayList;
import java.util.Iterator;

/**
 * Created by yuwen on 2017/7/3 0003.
 */
public class IrisClosureBlock {
    private IrisContextEnvironment m_currentEnvironment = null;

    private IrisObject m_nativeObject = null;

    private ArrayList<String> m_parameters = null;
    private String m_variableParameter = null;

    MethodHandle m_methodHandle = null;

    public IrisObject getNativeObject() {
        return m_nativeObject;
    }

    public void setNativeObject(IrisObject nativeObject) {
        m_nativeObject = nativeObject;
}

    public IrisClosureBlock(IrisContextEnvironment upperEnvironment, ArrayList<String> parameters, String variableParameter, MethodHandle methodHandle) {
        this.m_parameters = parameters;
        this.m_variableParameter = variableParameter;
        this.m_methodHandle = methodHandle;

        this.m_nativeObject = new IrisObject();
        this.m_nativeObject.setObjectClass(IrisDevUtil.GetClass("Block"));
        this.m_nativeObject.setNativeObject(this);

        this.m_currentEnvironment = CreateNewEnvironment(upperEnvironment);
    }

    private IrisContextEnvironment CreateNewEnvironment(IrisContextEnvironment upperEnvrionment) {

        IrisContextEnvironment newEnv = new IrisContextEnvironment();

        newEnv.setClosureBlockObj(m_nativeObject);
        newEnv.setRunTimeType(IrisContextEnvironment.RunTimeType.RunTime);
        newEnv.setUpperContext(upperEnvrionment);
        newEnv.setRunningType(null);

        return newEnv;
    }

    public IrisValue Call(ArrayList<IrisValue> parameters, IrisThreadInfo info) throws IrisExceptionBase {
        if(!ParameterCheck(parameters)) {
            throw new IrisParameterNotFitException(IrisDevUtil.GetCurrentThreadInfo().getCurrentFileName(),
                    IrisDevUtil.GetCurrentThreadInfo().getCurrentLineNumber(),
                    "Parameter not fit:" + parameters == null ? " 0" : " " + Integer.toString(parameters.size())
                            + " " +
                            "for " +
                            Integer.toString(m_parameters == null ? 0 : m_parameters.size()) + ".");
        }

        // add local variables
        if (parameters != null) {
//            for (int i = 0; i < parameters.size(); i++) {
//                m_currentEnvironment.AddLocalVariable(m_parameters.get(i), parameters.get(i));
//            }
            // parameter -> local variable && variable parameter process
            if (parameters != null && parameters.size() != 0) {
                int counter = 0;
                Iterator<IrisValue> iterator = parameters.iterator();
                for (String value : m_parameters) {
                    m_currentEnvironment.AddLocalVariable(value, iterator.next());
                    ++counter;
                }

                if (m_variableParameter != null) {
                    ArrayList<IrisValue> variables = new ArrayList<IrisValue>(parameters.subList(counter, parameters.size()));
                    IrisClass arrayClass = IrisDevUtil.GetClass("Array");
                    IrisValue arrayValue = arrayClass.CreateNewInstance(variables, null, info);
                    m_currentEnvironment.AddLocalVariable(m_variableParameter, arrayValue);
                }
            }
        }

        IrisValue result = null;
        try {
            result = (IrisValue)(m_methodHandle.invokeExact(m_currentEnvironment, info));
        } catch (Throwable throwable) {
            //throwable.printStackTrace();
            if(throwable instanceof IrisExceptionBase) {
                throw (IrisExceptionBase)throwable;
            }
            else {
                throwable.printStackTrace();
                throw new IrisUnkownFatalException("Unkown irregular happened.", info.getCurrentLineNumber(), info.getCurrentFileName());
            }
        }
        return result;
    }

    private boolean ParameterCheck(ArrayList<IrisValue> parameters) {
        if (parameters != null && parameters.size() > 0) {
            if (m_variableParameter != null) {
                return parameters.size() >= m_parameters.size();
            } else {
                return parameters.size() == m_parameters.size();
            }
        } else {
            return m_parameters == null || m_parameters.size() == 0;
        }
    }

    public IrisValue GetLocalVariable(String name) {
        IrisValue target = m_currentEnvironment.GetLocalVariableWithinChain(name);

        if(target == null) {
            target = IrisValue.CloneValue(IrisDevUtil.Nil());
            m_currentEnvironment.AddLocalVariable(name, target);
        }

        return target;
    }

    public void AddLocalVariable(String name, IrisValue value) {
        m_currentEnvironment.AddLocalVariable(name, value);
    }

    public IrisValue GetInstanceVariable(String name) {
        IrisContextEnvironment tmpEnv = m_currentEnvironment;
        IrisValue target = tmpEnv.GetLocalVariable(name);

        tmpEnv = tmpEnv.getUpperContext();
        while(tmpEnv != null && target == null) {
            if(tmpEnv.getRunningType() != null) {
                IrisObject obj = null;
                switch (tmpEnv.getRunTimeType()) {
                    case ClassDefineTime:
                        obj = ((IrisClass)(tmpEnv.getRunningType())).getClassObject();
                        break;
                    case ModuleDefineTime:
                        obj = ((IrisModule)(tmpEnv.getRunningType())).getModuleObject();
                        break;
//                    case InterfaceDefineTime:
//                        break;
                    case RunTime:
                        obj = (IrisObject)tmpEnv.getRunningType();
                        break;
                }
                target = obj.GetInstanceVariable(name);
                if(target == null) {
                    target = tmpEnv.GetLocalVariable(name);
                }
            }
            else {
                target = tmpEnv.GetLocalVariable(name);
            }
        }

        if(target == null) {
            target = IrisValue.CloneValue(IrisDevUtil.Nil());
            m_currentEnvironment.AddLocalVariable(name, target);
        }

        return target;
    }

    public IrisValue GetClassVariable(String name) {
        IrisContextEnvironment tmpEnv = m_currentEnvironment;
        IrisValue target = tmpEnv.GetLocalVariable(name);

        tmpEnv = tmpEnv.getUpperContext();
        while(tmpEnv != null && target == null) {
            if(tmpEnv.getRunningType() != null) {
                switch (tmpEnv.getRunTimeType()) {
                    case ClassDefineTime:
                        target = ((IrisClass)(tmpEnv.getRunningType())).GetClassVariable(name);
                        break;
                    case ModuleDefineTime:
                        target = ((IrisModule)(tmpEnv.getRunningType())).GetClassVariable(name);
                        break;
//                    case InterfaceDefineTime:
//                        break;
                    case RunTime:
                        target = ((IrisObject)tmpEnv.getRunningType()).getObjectClass().GetClassVariable(name);
                        break;
                }
                if(target == null) {
                    target = tmpEnv.GetLocalVariable(name);
                }
            }
            else {
                target = tmpEnv.GetLocalVariable(name);
            }

            tmpEnv = tmpEnv.getUpperContext();
        }

        if(target == null) {
            target = IrisValue.CloneValue(IrisDevUtil.Nil());
            m_currentEnvironment.AddLocalVariable(name, target);
        }

        return target;
    }

    public IrisValue GetConstance(String name) throws IrisExceptionBase {
        IrisContextEnvironment tmpEnv = m_currentEnvironment;
        IrisValue target = null;

        while(tmpEnv != null) {
            if(tmpEnv.getRunningType() != null) {
                switch (tmpEnv.getRunTimeType()) {
                    case ClassDefineTime:
                        target = ((IrisClass)(tmpEnv.getRunningType())).GetConstance(name);
                        break;
                    case ModuleDefineTime:
                        target = ((IrisModule)(tmpEnv.getRunningType())).GetConstance(name);
                        break;
//                    case InterfaceDefineTime:
//                        break;
                    case RunTime:
                        target = ((IrisObject)tmpEnv.getRunningType()).getObjectClass().GetConstance(name);
                        break;
                }
            }
            tmpEnv = tmpEnv.getUpperContext();
        }

        if(target == null) {
            target = IrisInterpreter.INSTANCE.GetConstance(name);
        }

        if(target == null) {
            throw new IrisConstanceNotFoundException(IrisDevUtil.GetCurrentThreadInfo().getCurrentFileName(),
                    IrisDevUtil.GetCurrentThreadInfo().getCurrentLineNumber(),
                    "Constance of " + name + " not found.");
        }

        return target;
    }

}

File: src/main/java/org/irislang/jiris/core/IrisClosureBlock.kt
package org.irislang.jiris.core

import org.irislang.jiris.IrisInterpreter
import org.irislang.jiris.core.exceptions.IrisExceptionBase
import org.irislang.jiris.core.exceptions.fatal.IrisConstanceNotFoundException
import org.irislang.jiris.core.exceptions.fatal.IrisParameterNotFitException
import org.irislang.jiris.core.exceptions.fatal.IrisUnkownFatalException
import org.irislang.jiris.dev.IrisDevUtil

import java.lang.invoke.MethodHandle
import java.util.ArrayList

/**
 * Created by yuwen on 2017/7/3 0003.
 */
class IrisClosureBlock(upperEnvironment: IrisContextEnvironment, parameters: ArrayList<String>,
                       variableParameter: String, methodHandle: MethodHandle) {
    private var m_currentEnvironment: IrisContextEnvironment

    var nativeObject: IrisObject

    private var m_parameters: ArrayList<String> = ArrayList()
    private var m_variableParameter: String? = null

    internal var m_methodHandle: MethodHandle

    init {
        this.m_parameters = parameters
        this.m_variableParameter = variableParameter
        this.m_methodHandle = methodHandle

        this.nativeObject = IrisObject()
        this.nativeObject.objectClass = IrisDevUtil.GetClass("Block")
        this.nativeObject.nativeObject = this

        this.m_currentEnvironment = CreateNewEnvironment(upperEnvironment)
    }

    private fun CreateNewEnvironment(upperEnvrionment: IrisContextEnvironment): IrisContextEnvironment {

        val newEnv = IrisContextEnvironment()

        newEnv.closureBlockObj = nativeObject
        newEnv.runTimeType = IrisContextEnvironment.RunTimeType.RunTime
        newEnv.upperContext = upperEnvrionment
        newEnv.runningType = null

        return newEnv
    }

    @Throws(IrisExceptionBase::class)
    fun Call(parameters: ArrayList<IrisValue>, info: IrisThreadInfo): IrisValue {
        if (!ParameterCheck(parameters)) {
            throw IrisParameterNotFitException(IrisDevUtil.GetCurrentThreadInfo().currentFileName,
                    IrisDevUtil.GetCurrentThreadInfo().currentLineNumber,
                    "Parameter not fit: ${parameters.size} for ${m_parameters.size}")
        }

            // parameter -> local variable && variable parameter process
        if (!parameters.isEmpty()) {
            var counter = 0
            val iterator = parameters.iterator()
            for (value in m_parameters) {
                m_currentEnvironment.AddLocalVariable(value, iterator.next())
                ++counter
            }

            if (m_variableParameter != null) {
                val variables = ArrayList(parameters.subList(counter, parameters.size))
                val arrayClass = IrisDevUtil.GetClass("Array")
                val arrayValue = arrayClass!!.CreateNewInstance(variables, this.m_currentEnvironment.upperContext!!, info)
                m_currentEnvironment.AddLocalVariable(m_variableParameter!!, arrayValue)
            }
        }

        val result: IrisValue
        try {
            result = m_methodHandle.invokeExact(m_currentEnvironment, info) as IrisValue
        } catch (throwable: Throwable) {
            if (throwable is IrisExceptionBase) {
                throw throwable
            } else {
                throwable.printStackTrace()
                throw IrisUnkownFatalException("Unkown irregular happened.", info.currentLineNumber, info.currentFileName)
            }
        }

        return result
    }

    private fun ParameterCheck(parameters: ArrayList<IrisValue>): Boolean {
        return if (parameters.size > 0) {
            if (m_variableParameter != null) {
                parameters.size >= m_parameters.size
            } else {
                parameters.size == m_parameters.size
            }
        } else {
            m_parameters.isEmpty()
        }
    }

    fun GetLocalVariable(name: String): IrisValue {
        var target = m_currentEnvironment.GetLocalVariableWithinChain(name)

        if (target == null) {
            target = IrisValue.CloneValue(IrisDevUtil.Nil())
            m_currentEnvironment.AddLocalVariable(name, target)
        }

        return target
    }

    fun AddLocalVariable(name: String, value: IrisValue) {
        m_currentEnvironment.AddLocalVariable(name, value)
    }

    fun GetInstanceVariable(name: String): IrisValue? {
        var tmpEnv: IrisContextEnvironment? = m_currentEnvironment
        var target: IrisValue? = tmpEnv!!.GetLocalVariable(name)

        tmpEnv = tmpEnv.upperContext
        while (tmpEnv != null && target == null) {
            if (tmpEnv.runningType != null) {
                var obj: IrisObject? = null
                when (tmpEnv.runTimeType) {
                    IrisContextEnvironment.RunTimeType.ClassDefineTime -> obj = (tmpEnv.runningType as IrisClass).classObject
                    IrisContextEnvironment.RunTimeType.ModuleDefineTime -> obj = (tmpEnv.runningType as IrisModule).moduleObject
                    IrisContextEnvironment.RunTimeType.RunTime -> obj = tmpEnv.runningType as IrisObject
                    IrisContextEnvironment.RunTimeType.InterfaceDefineTime -> TODO()
                }
                target = obj!!.GetInstanceVariable(name)
                if (target == null) {
                    target = tmpEnv.GetLocalVariable(name)
                }
            } else {
                target = tmpEnv.GetLocalVariable(name)
            }
        }

        if (target == null) {
            target = IrisValue.CloneValue(IrisDevUtil.Nil())
            m_currentEnvironment.AddLocalVariable(name, target)
        }

        return target
    }

    fun GetClassVariable(name: String): IrisValue? {
        var tmpEnv: IrisContextEnvironment? = m_currentEnvironment
        var target: IrisValue? = tmpEnv!!.GetLocalVariable(name)

        tmpEnv = tmpEnv.upperContext
        while (tmpEnv != null && target == null) {
            if (tmpEnv.runningType != null) {
                when (tmpEnv.runTimeType) {
                    IrisContextEnvironment.RunTimeType.ClassDefineTime -> target = (tmpEnv.runningType as IrisClass).GetClassVariable(name)
                    IrisContextEnvironment.RunTimeType.ModuleDefineTime -> target = (tmpEnv.runningType as IrisModule).GetClassVariable(name)
                    IrisContextEnvironment.RunTimeType.RunTime -> target = (tmpEnv.runningType as IrisObject).objectClass!!.GetClassVariable(name)
                    IrisContextEnvironment.RunTimeType.InterfaceDefineTime -> TODO()
                }
                if (target == null) {
                    target = tmpEnv.GetLocalVariable(name)
                }
            } else {
                target = tmpEnv.GetLocalVariable(name)
            }

            tmpEnv = tmpEnv.upperContext
        }

        if (target == null) {
            target = IrisValue.CloneValue(IrisDevUtil.Nil())
            m_currentEnvironment.AddLocalVariable(name, target)
        }

        return target
    }

    @Throws(IrisExceptionBase::class)
    fun GetConstance(name: String): IrisValue {
        var tmpEnv: IrisContextEnvironment? = m_currentEnvironment
        var target: IrisValue? = null

        while (tmpEnv != null) {
            if (tmpEnv.runningType != null) {
                when (tmpEnv.runTimeType) {
                    IrisContextEnvironment.RunTimeType.ClassDefineTime -> target = (tmpEnv.runningType as IrisClass).GetConstance(name)
                    IrisContextEnvironment.RunTimeType.ModuleDefineTime -> target = (tmpEnv.runningType as IrisModule).GetConstance(name)
                    IrisContextEnvironment.RunTimeType.RunTime -> target = (tmpEnv.runningType as IrisObject).objectClass!!.GetConstance(name)
                    IrisContextEnvironment.RunTimeType.InterfaceDefineTime -> TODO()
                }
            }
            tmpEnv = tmpEnv.upperContext
        }

        if (target == null) {
            target = IrisInterpreter.INSTANCE.GetConstance(name)
        }

        if (target == null) {
            throw IrisConstanceNotFoundException(IrisDevUtil.GetCurrentThreadInfo().currentFileName,
                    IrisDevUtil.GetCurrentThreadInfo().currentLineNumber,
                    "Constance of ${name} not found.")
        }

        return target
    }

}
